src/chainstate/stacks/db/blocks.rs

```diff
@@ -3889,6 +3889,49 @@ impl StacksChainState {
         Ok((fees, burns, receipts))
     }
 
+    /// If an epoch transition occurs at this Stacks block,
+    ///   apply the transition and return any receipts from the transition.
+    pub fn process_epoch_transition(
+        clarity_tx: &mut ClarityTx,
+        chain_tip_burn_header_height: u32,
+    ) -> Result<Vec<StacksTransactionReceipt>, Error> {
+        // is this stacks block the first of a new epoch?
+        let (stacks_parent_epoch, sortition_epoch) = clarity_tx.with_clarity_db_readonly(|db| {
+            (
+                db.get_clarity_epoch_version(),
+                db.get_stacks_epoch(chain_tip_burn_header_height),
+            )
+        });
+
+        if let Some(sortition_epoch) = sortition_epoch {
+            // the parent stacks block has a different epoch than what the Sortition DB
+            //  thinks should be in place.
+            if stacks_parent_epoch != sortition_epoch.epoch_id {
+                info!("Applying epoch transition"; "new_epoch_id" => %sortition_epoch.epoch_id, "old_epoch_id" => %stacks_parent_epoch);
+                // this assertion failing means that the _parent_ block was invalid: this is bad and should panic.
+                assert!(stacks_parent_epoch < sortition_epoch.epoch_id, "The SortitionDB believes the epoch is earlier than this Stacks block's parent: sortition db epoch = {}, parent epoch = {}", sortition_epoch.epoch_id, stacks_parent_epoch);
+                // time for special cases:
+                match stacks_parent_epoch {
+                    StacksEpochId::Epoch10 => {
+                        panic!("Clarity VM believes it was running in 1.0: pre-Clarity.")
+                    }
+                    StacksEpochId::Epoch20 => {
+                        assert_eq!(
+                            sortition_epoch.epoch_id,
+                            StacksEpochId::Epoch21,
+                            "Should only transition from Epoch20 to Epoch21"
+                        );
+                        clarity_tx.block.initialize_epoch_2_1()?;
+                    }
+                    StacksEpochId::Epoch21 => {
+                        panic!("No defined transition from Epoch21 forward")
+                    }
+                }
+            }
+        }
+        Ok(vec![])
```

Comment:
Are there going to be receipts from this?

Reply:
Ah, yes, there should be.

Reply:
Added in d6458eccc939ccf7806470012568b90a5470fda9

Reply:
The receipt from `initialize_epoch_2_1()` needs to be plumbed through here.

Reply:
Fixed in 92caa6d